agent/consul/autopilot/autopilot.go

```diff
@@ -0,0 +1,457 @@
+	a.waitGroup.Wait()
+}
+
+// autopilotLoop periodically looks for nonvoting servers to promote and dead servers to remove.
+func (a *Autopilot) run() {
+	defer a.waitGroup.Done()
+
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(a.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-a.shutdownCh:
+			return
+		case <-ticker.C:
+			autopilotConfig, ok := a.delegate.GetOrCreateAutopilotConfig()
+			if !ok {
+				continue
+			}
+
+			// Skip the non-voter promotions unless all servers support the new APIs
+			minRaftProtocol, err := a.MinRaftProtocol()
+			if err != nil {
+				a.logger.Printf("[ERR] autopilot: error getting server raft protocol versions: %s", err)
+				continue
+			}
+			if minRaftProtocol >= 3 {
+				promotions, err := a.delegate.PromoteNonVoters(autopilotConfig, a.GetClusterHealth())
+				if err != nil {
+					a.logger.Printf("[ERR] autopilot: Error checking for non-voters to promote: %s", err)
+				}
+				if err := a.handlePromotions(promotions); err != nil {
+					a.logger.Printf("[ERR] autopilot: Error handling promotions: %s", err)
+				}
+			}
+
+			if err := a.pruneDeadServers(autopilotConfig); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error checking for dead servers to remove: %s", err)
+			}
+		case <-a.removeDeadCh:
+			autopilotConfig, ok := a.delegate.GetOrCreateAutopilotConfig()
+			if !ok {
+				continue
+			}
+
+			if err := a.pruneDeadServers(autopilotConfig); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error checking for dead servers to remove: %s", err)
+			}
+		}
+	}
+}
+
+// fmtServer prints info about a server in a standard way for logging.
+func fmtServer(server raft.Server) string {
+	return fmt.Sprintf("Server (ID: %q Address: %q)", server.ID, server.Address)
+}
+
+// pruneDeadServers removes up to numPeers/2 failed servers
+func (a *Autopilot) pruneDeadServers(conf *Config) error {
+	if !conf.CleanupDeadServers {
+		return nil
+	}
+
+	// Failed servers are known to Serf and marked failed, and stale servers
+	// are known to Raft but not Serf.
+	var failed []string
+	staleRaftServers := make(map[string]raft.Server)
+	raftNode := a.delegate.Raft()
+	future := raftNode.GetConfiguration()
+	if err := future.Error(); err != nil {
+		return err
+	}
+	for _, server := range future.Configuration().Servers {
+		staleRaftServers[string(server.Address)] = server
+	}
+	serfLAN := a.delegate.Serf()
+	for _, member := range serfLAN.Members() {
+		valid, parts := metadata.IsConsulServer(member)
+		if valid {
+			if _, ok := staleRaftServers[parts.Addr.String()]; ok {
+				delete(staleRaftServers, parts.Addr.String())
+			}
+
+			if member.Status == serf.StatusFailed {
+				failed = append(failed, member.Name)
+			}
+		}
+	}
+
+	// We can bail early if there's nothing to do.
+	removalCount := len(failed) + len(staleRaftServers)
+	if removalCount == 0 {
+		return nil
+	}
+
+	// Only do removals if a minority of servers will be affected.
+	peers, err := a.delegate.NumPeers()
```

Comment:
This is super critical to get right since we need to count the voters (the old Consul version of `numPeers()` had a subtle bug around this. Since this count is based off the Raft configuration, I'd move that logic into this autopilot package via a helper function that you pass the `future.Configuration()` into here and counts the voters. This will have the benefit of making sure we have the same logic in both products, it will save us another Raft configuration fetch, and it will avoid bugs where the configuration might have changed between fetches.

This'll let us remove `NumPeers` from the delegate, too, simplifying the interface.